Fix part of #5914: UserStatsModel Takeout

ulloaluis opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes part of issue #5914. Implements Takeout functionality on UserStatsModel in core/storage/user/gae_models.py. Includes appropriate backend testing in core/storage/user/gae_models_test.py.

**Write-up:**
We will export the five properties relevant to the UserStatsModel. As Ben (Henning) suggested, we should include any data we want to potentially give to the user. None of the five properties reveal sensitive data about the user, so they are all safe to be exported. The five properties we are exporting are all useful and possibly wanted statistics related to the popularity of explorations the user has contributed to/owned*. We are not exporting the schema_version since that is not relevant to the user.

**Properties from UserStatsModel:**
- impact_score
- total_plays
- average_ratings
- num_ratings
- weekly_creator_stats_list

The _impact score_ (float) for a particular user (exact function defined in UserStatsModel documentation) is a metric calculated over the explorations for which a user is a contributor. A higher impact score is indicative of how well those explorations do, based on ratings.

The _total_plays_ (integer), _average_ratings_ (float), and _num_ratings_ (integer) variables are based on all the explorations owned by a user. total_plays is the sum of all the plays on those explorations and average_ratings is the average of all the ratings on those explorations. num_ratings is simply the number of ratings left in total across all of those (user-owned) explorations.

The _weekly_creator_stats_list_ is a list which stores JSON objects. Each JSON object contains the user’s average_ratings and total_plays for the specified date they are key’d on. See documentation of core/storage/user/gae_models.py UserStatsModel for more detail an example.

**Expected export structure** (This model will not need any special exporting from takeout_processor_service.py, since export_data already exports the data in the specified format.)
```
creator_impact_metrics:
    impact_score: <float_impact_score>
    total_plays: <integer_total_plays>
    average_ratings: <float_average_ratings>
    num_ratings: <integer_num_ratings>
    weekly_creator_stats_list: 
       - date_1: <datetime_when_following_stats_were_collected>
          average_ratings: <float_average_ratings>
          total_plays: <integer_total_plays>
       - date_2: <datetime_when_following_stats_were_collected>
          average_ratings: <float_average_ratings>
          total_plays: <integer_total_plays>
          ...
```
**

*Note that only impact score is based on explorations the user has contributed to. The rest are based specifically on explorations the user owns (according to documentation).
**The “ - “ here signifies a JSON object of the form
```
{
    date_k: {
        average_ratings: <float_average_ratings>
        total_plays: <integer_total_plays>
    }
}
```
where _date_k_ itself a python datetime specifying when the following _average_ratings_ and _total_plays_ metrics were recorded.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

ulloaluis committed:
initial changes

ulloaluis committed:
misc. bug resolved

ulloaluis commented:
PTAL @seanlip possibly also @BenHenning

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -384,6 +384,28 @@ def get_or_create(cls, user_id):
+            dict. The user-relevant properties of UserStatsModel in a python
+                dict format.
+        """
+        user_model = UserStatsModel.get_or_create(user_id)
```
This should not be creating, I think. I think it's OK to return None in this case.

/cc @BenHenning what do you think?

ulloaluis replied:
bumping this @BenHenning @seanlip

BenHenning replied:
This should probably be a hard failure, though we can either do it here or at a higher level. It probably is clearer to do it here, though. Let's raise an exception if the user isn't found, and make it clear in the documentation that the caller is responsible for ensure the user ID passed in here corresponds to a real user.

BenHenning replied:
After following up with Sean, in this case we should probably return None since it's valid for a real user not to have data in this model.

For other models, we should decide between throwing an error and returning None based on the same thing: is it valid for the user not to have data in that model?

ulloaluis replied:
addressed in latest commit

seanlip reviewed (changes requested):
Hi @ulloaluis LGTM, thanks! Just one thought.

@BenHenning would you mind taking a quick look too, just as a sanity-check?

BenHenning commented on core/storage/user/gae_models_test.py:
```diff
@@ -84,6 +84,69 @@ def test_get_failure(self):
+            self.USER_1_WEEKLY_CREATOR_STATS_LIST)
+        user_models.UserStatsModel.put(user_model_1)
+
+    def test_export_data_on_existing_user(self):
```
Maybe also add a multi-user test, e.g. that if you retrieve the data for a different user that it's the data for that user rather than the first user.

ulloaluis replied:
addressed in latest commit

BenHenning reviewed (changes requested):
Overall LGTM, but would like to do one final pass after addressing the suggested changes.

BenHenning commented on core/storage/user/gae_models.py:
```diff
@@ -384,6 +384,28 @@ def get_or_create(cls, user_id):
+            'total_plays': user_model.total_plays,
+            'average_ratings': user_model.average_ratings,
+            'num_ratings': user_model.num_ratings,
+            'weekly_creator_stats_list': user_model.weekly_creator_stats_list
```
Per https://docs.google.com/document/d/1WSJYujTTGUADFakMD_-cpAuWZI4yWakL7cXOT3DMx7w/edit#bookmark=id.xfw7p3frtru8 I see that we're following the same format roughly, but I suggest we actually explicitly assemble the weekly_creator_stats_list to guard against accidentally pulling in changes to the underlying structure that we might not want to export to users.

WDYT?

ulloaluis replied:
I'm okay with explicitly assembling the weekly_creator_stats_list. There might be a varying amount of items in that list depending on when the user started creating/contributing to explorations, so I wanted to confirm what I would be doing differently in manually constructing the list. In constructing the list, should I be verifying that each object is of the expected form? So verifying types with **type(obj)** and double checking key-names (as it adheres to https://docs.google.com/document/d/1WSJYujTTGUADFakMD_-cpAuWZI4yWakL7cXOT3DMx7w/edit?ts=5c89a9c3#). What should I do if the list contains invalid data? Raise an error or simply not include that item when manually creating the list? Thanks. @BenHenning PTAL

BenHenning replied:
I think raising an error is fine--if Takeout fails due to such an error, we should investigate it. Hopefully our future higher-level Takeout tests help prevent these types of situations, but failing here makes sense.

I didn't think about doing the type validation, and I'm not convinced we necessarily need it. I was more concerned with what labeled data we're including, not the contents of that data. We can't verify everything, but if we trust the labels to be correct, then manually reassembling the whole structure makes it really clear what we're choosing to include in Takeout data. Specifically, I see Takeout as an opt-in mechanism for data exporting, not opt-out. Maybe @seanlip has a different view here.

I also added one more comment hoping to clarify one of the bits of information we're exporting here.

seanlip replied:
What @BenHenning says sounds fine to me in all respects (thanks Ben)!

ulloaluis committed:
Verifying weekly_creator_stats_list, testing export_data for multiple users, testing that export_data returns None when user_id not in data store.

ulloaluis commented:
Addressed comments in latest commit @seanlip @BenHenning. Note that I was attempting to keep python2/3 compatibility when creating the verified_weekly_creator_stats_list, so I was trying not to use things like iteritems(). PTAL

ulloaluis commented:
@BenHenning PTAL

BenHenning commented on core/storage/user/gae_models.py:
```diff
@@ -384,6 +384,41 @@ def get_or_create(cls, user_id):
+                if isinstance(weekly_stat[key_obj], dict):
+                    key_entries = sorted(list(weekly_stat[key_obj].keys()))
+                    if key_entries == ['average_ratings', 'total_plays']:
+                        validated_weekly_creator_stats_list.append(weekly_stat)
```
I'm not sure this fully fixes the issue--shouldn't the weekly_stat structure also be explicitly defined here to guard against its structure possibly changing in the future?

BenHenning reviewed (approved):
Thanks! I think it LGTM, but I think we should expland the weekly_stat structure if it's actually an object like it seems to be. If it's just a number, then feel free to ignore my comment.

If you choose to remove the type check validation and throw errors, please FYI me and I'll take another look.

ulloaluis commented:
Note: Academic quarter just ended, so I'm in transit between school and an internship, but I should have time to finish this PR in the next 1-3 days (posted to remove stale label).

ulloaluis committed:
Now explicitly constructing list with a list comprehension without error checking. Error checking is moved to testing file in the form of type/content validation for weekly_creator_stats_list.

ulloaluis commented:
Changes:
1) Now explicitly constructing weekly_creator_stats_list with a list comprehension with error checking removed. This improves speed/efficiency (important for the user) and makes logical sense since we would want to catch issues with malformed weekly_creator_stats_list objects in the testing state (so that they won't be in production at all when exporting user data). In particular, if another developer modifies the weekly_creator_stats_list structure, a test in this file will fail.
2) Added test that explicitly checks and confirms the contents and types of the objects in the weekly_creator_stats_list. The expected format of the weekly_creator_stats_list is in the **Explanation** for this PR.

ulloaluis commented:
PTAL @BenHenning @seanlip

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -384,6 +384,35 @@ def get_or_create(cls, user_id):
+            return None
+
+        weekly_stats = user_model.weekly_creator_stats_list
+        weekly_stats_constructed = [weekly_stat for weekly_stat in weekly_stats]
```
You need to break down each weekly_stat dict explicitly, to its finest-grained details -- the full list of fields and subfields in every returned Takeout dict should be manually assembled. (I thought we talked about this; see @BenHenning's earlier comment, too.)

ulloaluis replied:
Addressed.

seanlip commented on core/storage/user/gae_models_test.py:
```diff
@@ -84,6 +84,127 @@ def test_get_failure(self):
+        self.assertEqual(user_data, test_data)
+
+    def test_export_data_stats_list_is_well_formed(self):
+        """Test if weekly creator stats list has only the expected elements."""
```
Isn't this test essentially a subset of test_export_data_on_existing_user()? Why do we need it / what information does it give us that that other test does not?

ulloaluis replied:
PTAL @seanlip  at the updated version (one line changed).

**What I was getting at**: export_data explicitly constructs the stats list*, so the purpose of this test was to make sure that the weekly_creator_stats_list in the class (not the export_data) contains the expected elements. Otherwise, it's possible that we did not fully export all the data in the stats list since someone decided to add more entries to the stats list (entries like 'total_plays'). This test will catch that change and prompt whoever is adding more data to git-blame the file and get in contact with me or someone else so we can properly update export_data to include their new features. 

**What was wrong**: I used export_data, not the weekly_creator_stats_list of the user model from the data store. I'll fix that--my bad.

Is this not necessary? Why?

*latest PR should do it correctly

ulloaluis replied:
Is this test fine given the recent developments about how we are approaching takeout (timeline wise), or should it be dropped/modified? @seanlip

seanlip replied:
I think it can be dropped. Would you agree?

Also, thanks for checking!

ulloaluis replied:
Agreed!

seanlip reviewed (changes requested):
@ulloaluis I left some comments, PTAL.

@varun-tandon Could you please review the next iterations of @ulloaluis's PRs?

ulloaluis committed:
fixed explicit construction of weeklycreatorstatslist and revamped test

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -384,6 +384,45 @@ def get_or_create(cls, user_id):
+        weekly_stats = user_model.weekly_creator_stats_list
+        weekly_stats_constructed = []
+        for weekly_stat in weekly_stats:
+            date_key = weekly_stat.keys()[0]
```
Should you be iterating through the keys instead of just taking the first one? The [0] looks weird...

ulloaluis replied:
Each element in the list looks like this:
```
{
    date_k: {
        average_ratings: <float_average_ratings>
        total_plays: <integer_total_plays>
    }
}
```

(each obj is a dict with one key that maps to another dict.)

varun-tandon replied:
This looks good to me. @seanlip does Luis' explanation suffice for you?

seanlip replied:
I think it's better to do this in generality. Can you iterate through the keys and then do a dict comprehension in the constructed_stat, in such a way that it would work if there were multiple keys in the dict?

ulloaluis replied:
Okay, I think this is addressed now.

ulloaluis committed:
Addressed accessing first element issue

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -384,6 +384,45 @@ def get_or_create(cls, user_id):
+        weekly_stats = user_model.weekly_creator_stats_list
+        weekly_stats_constructed = []
+        for weekly_stat in weekly_stats:
+            for date_key in weekly_stat.keys():
```
You don't need the ".keys()" part.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -384,6 +384,45 @@ def get_or_create(cls, user_id):
+            for date_key in weekly_stat.keys():
+                stat_dict = weekly_stat[date_key]
+                constructed_stat = {
+                    (date_key): {
```
You don't need the parens around date_key.

ulloaluis committed:
removed parens, removed .keys(), removed a test

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -384,6 +384,45 @@ def get_or_create(cls, user_id):
+                dict format.
+        """
+        user_model = UserStatsModel.get(user_id, strict=False)
+        if not user_model:
```
Can you test this code path as well please?

seanlip replied:
Oh wait ignore that, you've already done so. LGTM!

seanlip reviewed (commented):
Sorry @ulloaluis, one more thing I forgot (I noticed it after reviewing the other PR)

seanlip reviewed (approved):
Whoops, my bad. LGTM. Thanks @ulloaluis!

ulloaluis commented:
Hmm. Any idea why this failed a few E2E tests? @seanlip

seanlip commented:
Not sure ... I tried restarting, just in case -- let's see if that works.

ulloaluis commented:
Looks like it was #7055. Seems to be resolved now @seanlip

seanlip commented:
Great! Thanks @ulloaluis, merging :)
